### PR TITLE
feat: add total users insight

### PR DESCRIPTION
**Pull request by BrunoQuaresma**
Related to https://github.com/coder/coder/issues/15297

### CODE REVIEW

File: coderd/insights.go [hunk 1]
```diff
@@ -470,6 +470,70 @@ func (api *API) insightsTemplates(rw http.ResponseWriter, r *http.Request) {
...
+	)
+	p.ErrorExcessParams(vals)
+	if len(p.Errors) > 0 {
+		httpapi.Write(ctx, rw, http.StatusBadRequest, codersdk.Response{
+			Message:     "Query parameters have invalid values.",
+			Validations: p.Errors,
+		})
+		return
+	}
+
```
Comment (defelmnq) [inline coderd/insights.go:503]: Do we want to add validation for the `start_time` being set to a value after `end_time` ?

File: coderd/database/queries/insights.sql [hunk 2]
```diff
@@ -771,3 +771,18 @@ SELECT
...
+-- in the given timeframe. It returns the accumulated number of users for each date
+-- within the specified timeframe, providing a running total of user sign-ups.
+SELECT
+	date_trunc('day', created_at)::date AS date,
+	COUNT(*) OVER (ORDER BY date_trunc('day', created_at)::date) AS total
+FROM
+	users
+WHERE
+	created_at >= @start_time::timestamptz
+	AND created_at < @end_time::timestamptz
```
Comment (defelmnq) [inline coderd/database/queries/insights.sql:786]: More for confirmation than a request to change - is it expected to exclude the end_time from the query ? 

If I say I want the insights from 01/01to 02/01- should we include 02/01 ?

**Review by mtojek — COMMENTED** [review]
CI is failing at the moment, you need to take a look at the report and address issues.

File: coderd/database/dbmem/dbmem.go [hunk 3]
```diff
@@ -2320,6 +2320,48 @@ func (q *FakeQuerier) GetAPIKeysLastUsedAfter(_ context.Context, after time.Time
...
+func (q *FakeQuerier) GetAccumulatedUsersInsights(_ context.Context, arg database.GetAccumulatedUsersInsightsParams) ([]database.GetAccumulatedUsersInsightsRow, error) {
+	q.mutex.RLock()
+	defer q.mutex.RUnlock()
+
+	newUsersByDate := make(map[time.Time]int64)
+	for _, user := range q.users {
+		if user.CreatedAt.Before(arg.StartTime) || user.CreatedAt.After(arg.EndTime) {
+			continue
+		}
+		date := user.CreatedAt.Truncate(24 * time.Hour)
```
Comment (mafredri) [inline coderd/database/dbmem/dbmem.go:2332]: Will this work correctly with timezones?

File: coderd/database/queries/insights.sql (hunk 2 above)
Comment (mafredri) [reply coderd/database/queries/insights.sql:786]: In-memory DB doesn't match this:

```go
		if user.CreatedAt.Before(arg.StartTime) || user.CreatedAt.After(arg.EndTime) {
```

Missing `|| user.CreatedAt.Equal(arg.EndTime)`.

File: coderd/database/queries/insights.sql [hunk 4]
```diff
@@ -771,3 +771,18 @@ SELECT
 FROM unique_template_params utp
 JOIN workspace_build_parameters wbp ON (utp.workspace_build_ids @> ARRAY[wbp.workspace_build_id] AND utp.name = wbp.name)
 GROUP BY utp.num, utp.template_ids, utp.name, utp.type, utp.display_name, utp.description, utp.options, wbp.value;
+
+-- name: GetAccumulatedUsersInsights :many
+-- GetAccumulatedUsersInsights returns the accumulated number of users created
+-- in the given timeframe. It returns the accumulated number of users for each date
+-- within the specified timeframe, providing a running total of user sign-ups.
+SELECT
+	date_trunc('day', created_at)::date AS date,
```
Comment (mafredri) [inline coderd/database/queries/insights.sql:780]: This might produce strange results if requested start time and end time are in a different timezone. Each day may be missing few/many entries and we might get ghost dates as well (+/-1 day outside start/end range).

Suggestion:
```sql
	date_trunc('day', (
		created_at AT TIME ZONE EXTRACT(TIMEZONE FROM @start_time::timestamptz)
	)) AT TIME ZONE EXTRACT(TIMEZONE FROM @start_time::timestamptz) AS date,
```

(Untested), why the double at timezone? One strips timezone and the other one adds it back.

```
coder=# select created_at from users;
          created_at
-------------------------------
 2024-04-08 11:41:35.096834+00
 2024-04-08 11:40:54.587231+00
(2 rows)

coder=# select created_at at time zone 'europe/helsinki' from users;
          timezone
----------------------------
 2024-04-08 14:41:35.096834
 2024-04-08 14:40:54.587231
(2 rows)

coder=# select created_at at time zone 'europe/helsinki' at time zone 'europe/helsinki' from users;
           timezone
-------------------------------
 2024-04-08 11:41:35.096834+00
 2024-04-08 11:40:54.587231+00
(2 rows)
```

Notice +00 and lack thereof. Now notice how we get a timezoneally correct truncation when we do it without timezone:

```
coder=# select date_trunc('day', created_at) from users;
       date_trunc
------------------------
 2024-04-08 00:00:00+00
 2024-04-08 00:00:00+00
(2 rows)

coder=# select date_trunc('day', created_at at time zone 'europe/helsinki') at time zone 'europe/helsinki' from users;
        timezone
------------------------
 2024-04-07 21:00:00+00
 2024-04-07 21:00:00+00
(2 rows)
```

You might want to use a subquery so you can count over the improved truncation without repeating all that noise.

File: coderd/insights_test.go [hunk 5]
```diff
@@ -2351,3 +2351,101 @@ func TestGenericInsights_RBAC(t *testing.T) {
...
+		require.Equal(t, res[2].Date, today().Format(time.DateOnly))
+		require.Equal(t, res[2].Total, uint64(2))
+	})
+}
+
+func daysAgo(days int) time.Time {
+	return time.Now().UTC().AddDate(0, 0, -days).Truncate(24 * time.Hour)
+}
+
+func today() time.Time {
```
Comment (mafredri) [inline coderd/insights_test.go:2449]: This function name is a bit confusing. I know it makes sense for the test but might be better to define it inside the parent test function instead of entire `coderd_test` package.

File: coderd/insights_test.go [hunk 6]
```diff
@@ -2351,3 +2351,101 @@ func TestGenericInsights_RBAC(t *testing.T) {
...
+		dbgen.User(t, db, database.User{
+			Email:     "user3@coder.com",
+			Username:  "user3",
+			CreatedAt: daysAgo(1),
+		})
+
+		// When: requesting the accumulated total of users by day
+		res, err := client.TotalUsersInsight(ctx, codersdk.TotalUsersInsightRequest{
+			StartTime: daysAgo(2),
+			EndTime:   today(),
```
Comment (mafredri) [inline coderd/insights_test.go:2387]: It would be good to add tests with other timezones too. Might need to consider the value of user created at more closely to trigger edge cases.

File: coderd/insights.go [hunk 7]
```diff
@@ -470,6 +470,70 @@ func (api *API) insightsTemplates(rw http.ResponseWriter, r *http.Request) {
...
+		httpapi.Write(ctx, rw, http.StatusInternalServerError, codersdk.Response{
+			Message: "Internal error fetching total users.",
+			Detail:  err.Error(),
+		})
+		return
+	}
+
+	var res codersdk.TotalUsersInsightResponse
+	currentTotal := uint64(0)
+	for d := startTime; d.Before(endTime) || d.Equal(endTime); d = d.AddDate(0, 0, 1) {
```
Comment (mafredri) [inline coderd/insights.go:523]: Tip: One alternative to this method is to generate the empty dates as well via `generate_series` in the query. But don't feel you need to make the change now, just suggesting it as an option for the future.

**Review by mafredri — COMMENTED** [review]
Nice work! Had a few concern regarding timezones but otherwise this looks solid.

File: coderd/database/queries/insights.sql (hunk 2 above)
Comment (BrunoQuaresma) [reply coderd/database/queries/insights.sql:786]: @defelmnq I will double-check that.

@mafredri I didn't get your last comment. Sorry 😞

File: coderd/insights.go (hunk 7 above)
Comment (BrunoQuaresma) [reply coderd/insights.go:523]: I will take a look at that.

**Comment by BrunoQuaresma** [comment]
@mafredri About the timezone concern. I don't see us doing extra logic to handle timezones in other insights queries, why for this one should we care?

File: coderd/testdata/insights/template/parameters_two_dates_ago,_no_data.json.golden
Comment (BrunoQuaresma) [inline coderd/testdata/insights/template/parameters_two_dates_ago,_no_data.json.golden:1]: @mafredri not sure why this golden file was not added before 🤔

File: coderd/insights.go [hunk 8]
```diff
@@ -470,6 +470,67 @@ func (api *API) insightsTemplates(rw http.ResponseWriter, r *http.Request) {
...
+		})
+		return
+	}
+
+	startTime, endTime, ok := parseInsightsStartAndEndTime(ctx, rw, time.Now(), startTimeString, endTimeString)
+	if !ok {
+		return
+	}
+
+	zone, _ := startTime.Zone()
```
Comment (mafredri) [inline coderd/insights.go:509]: You might find this in your tests too but Zone it's not guaranteed to return a name. It's unfortunately safer to use the offset (integer) and turn it into a string like `2 => "2"` and `-2 => "-2".

That's a bit unfortunate however since there may be a minor edge case during DST shift and could potentially produce an unwanted extra/missing row depending on the data. I haven't tried to verify if/how much of a problem it would be though.

Comment (BrunoQuaresma) [reply coderd/insights.go]: Changed to use offset!

File: coderd/insights.go [hunk 9]
```diff
@@ -470,6 +470,67 @@ func (api *API) insightsTemplates(rw http.ResponseWriter, r *http.Request) {
...
+	startTime, endTime, ok := parseInsightsStartAndEndTime(ctx, rw, time.Now(), startTimeString, endTimeString)
+	if !ok {
+		return
+	}
+
+	_, offset := startTime.Zone()
+	rows, err := api.Database.GetAccumulatedUsersInsights(ctx, database.GetAccumulatedUsersInsightsParams{
+		StartTime: startTime,
+		EndTime:   endTime,
+		Timezone:  string(offset),
```
Comment (mafredri) [inline coderd/insights.go:513]: This doesn't behave like intended. You need to use the strconv package instead.

**Comment by BrunoQuaresma** [comment]
@mafredri I need your help here.

For the `Success` test, I'm creating four users + the first user created to be used with the client so the total of users should be five. 

When the tests run in memory, they return five as expected but when running with a PG database they return four. I think this is related to how dates are casting in the query but I have no idea how to fix it and I would love a second pair of eyes on it.

**Comment by SasSwart** [comment]
Note:
https://github.com/coder/coder/issues/15297#issuecomment-2500900916
https://github.com/coder/coder/issues/15297#issuecomment-2500903383

I'm removing `GetAccumulatedUsersInsights` from this PR and reducing its scope to the frontend changes to existing graphs. The new gauge required in https://github.com/coder/coder/issues/15297 will follow in a subsequent PR.

**Comment by SasSwart** [comment]
Looking into how entitlements count license utilization showed that instead of `GetAccumulatedUsersInsights` we'll need to use `GetActiveUserCount` for the gauge. This is good news, makes the PR smaller and means the number won't drift.

File: coderd/workspaces.go [hunk 10]
```diff
@@ -517,6 +517,8 @@ func createWorkspace(
 	// Update audit log's organization
 	auditReq.UpdateOrganizationID(template.OrganizationID)
 
+	// TODO (SasSwart): api.Authorize below is already done above. This seems redundant.
+	// 		Should we remove this?
```
Comment (johnstcn) [inline coderd/workspaces.go:521]: Yep, that does look redundant. Let's remove it in a separate PR.

**Comment by SasSwart** [comment]
closing in favor of https://github.com/coder/coder/pull/15683
